Content.Client/_Sunrise/Drugs/RainbowSmoothOverlay.cs

```diff
@@ -0,0 +1,191 @@
+    private StatusEffectsSystem _status = default!;
+
+	private RainbowSmoothOverlay? _overlay;
+
+	public override void Initialize()
+	{
+		base.Initialize();
+
+		SubscribeLocalEvent<SeeingRainbowsWeakStatusEffectComponent, StatusEffectAppliedEvent>(OnApplied);
+		SubscribeLocalEvent<SeeingRainbowsWeakStatusEffectComponent, StatusEffectRemovedEvent>(OnRemoved);
+		SubscribeLocalEvent<SeeingRainbowsWeakStatusEffectComponent, StatusEffectRelayedEvent<LocalPlayerAttachedEvent>>(OnPlayerAttached);
+		SubscribeLocalEvent<SeeingRainbowsWeakStatusEffectComponent, StatusEffectRelayedEvent<LocalPlayerDetachedEvent>>(OnPlayerDetached);
+		_overlay = new();
+        _status = _sysMan.GetEntitySystem<StatusEffectsSystem>();
+	}
+
+	public override void Shutdown()
+	{
+		base.Shutdown();
+        if (_overlay != null)
+        {
+            _overlayMan.RemoveOverlay(_overlay);
+            _overlay = null;
+        }
+	}
+
+	private void OnRemoved(Entity<SeeingRainbowsWeakStatusEffectComponent> ent, ref StatusEffectRemovedEvent args)
+	{
+		if (_player.LocalEntity != args.Target)
+			return;
+		if (_overlay != null)
+		{
+			_overlay.Intoxication = 0;
+			_overlay.TimeTicker = 0;
+			_overlay.CachedEndTime = null;
+			_overlayMan.RemoveOverlay(_overlay);
+		}
+	}
+
+	private void OnApplied(Entity<SeeingRainbowsWeakStatusEffectComponent> ent, ref StatusEffectAppliedEvent args)
+	{
+		if (_player.LocalEntity != args.Target)
+			return;
+		if (_overlay == null)
+			_overlay = new();
+		_overlay.Phase = _random.NextFloat(MathF.Tau);
+        // Кэшируем реальное время окончания эффекта
+        if (_status.TryGetEffectsEndTimeWithComp<SeeingRainbowsWeakStatusEffectComponent>(args.Target, out var endTime))
+            _overlay.CachedEndTime = endTime ?? TimeSpan.MaxValue;
+        else
+            _overlay.CachedEndTime = _timing.CurTime + TimeSpan.FromSeconds(60.0f); // fallback
+		_overlayMan.AddOverlay(_overlay);
+	}
+
+	private void OnPlayerAttached(Entity<SeeingRainbowsWeakStatusEffectComponent> ent, ref StatusEffectRelayedEvent<LocalPlayerAttachedEvent> args)
+	{
+        // Защита от ретрансляции событий нелокальной цели
+        if (_player.LocalEntity != args.Args.Entity)
+            return;
+        if (_overlay == null)
+            _overlay = new();
+        _overlayMan.AddOverlay(_overlay);
+        // Восстановить CachedEndTime если возможно
+        if (_status.TryGetEffectsEndTimeWithComp<SeeingRainbowsWeakStatusEffectComponent>(args.Args.Entity, out var endTime))
+            _overlay.CachedEndTime = endTime ?? TimeSpan.MaxValue;
+	}
+
+	private void OnPlayerDetached(Entity<SeeingRainbowsWeakStatusEffectComponent> ent, ref StatusEffectRelayedEvent<LocalPlayerDetachedEvent> args)
+	{
+        if (_player.LocalEntity != args.Args.Entity)
+            return;
+        if (_overlay != null)
+        {
+            _overlay.Intoxication = 0;
+            _overlay.TimeTicker = 0;
+            _overlayMan.RemoveOverlay(_overlay);
+        }
```

Comment:
не принято так делать
```suggestion
        if (_overlay == null)
            return;
         
        _overlay.Intoxication = 0;
        _overlay.TimeTicker = 0;
        _overlayMan.RemoveOverlay(_overlay);
```